tests/baselines/reference/intraExpressionInferencesReverseMappedTypes.errors.txt

```diff
@@ -0,0 +1,148 @@
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(67,21): error TS18046: 'x' is of type 'unknown'.
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(71,21): error TS18046: 'x' is of type 'unknown'.
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(80,21): error TS18046: 'x' is of type 'unknown'.
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(86,21): error TS18046: 'x' is of type 'unknown'.
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(95,21): error TS18046: 'x' is of type 'unknown'.
+tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferencesReverseMappedTypes.ts(101,21): error TS18046: 'x' is of type 'unknown'.

```

Comment:
All of those relate to the tuple-based tests that I added and those are **not** expected. 

It doesn't work because in tuples both `T` and `T[K]` are **not** reverse mapped symbols/types and thus the `inferReverseMappedType` isn't called from within `checkExpressionWithContextualType` (and only within that call `intraExpressionInferenceSites` and "pushed inference contexts" are available).

With objects `createReverseMappedType` only creates a `ReverseMapped` type and doesn't do much else. `inferReverseMappedType` gets called for properties later from `getTypeOfReverseMappedSymbol`). However, with tuples it calls `inferReverseMappedType` eagerly:
https://github.dev/microsoft/TypeScript/blob/eb014a26522dd809ae4d0e85634a62eabda2755a/src/compiler/checker.ts#L24090-L24101

I could try to fix this in this PR but I also feel like this would require changes that could be moved to a separate followup PR. This PR could focus on the base mechanism for this improvement and on object-based cases.